### DISCUSSION TITLE
"stable" ABI?

### DISCUSSION THREAD

**Discussion by James-E-A** [pypy/pypy#4884] [Q&A] [answered]
I'm building a package, and I want to include a precompiled wheel for PyPy, but it's my *understanding* that `py3-abi3-{tag.platform}` wheels aren't supported by PyPy.

If my understanding's correct, and the so-called `abi3` is *actually* CPython-specific, then I ask: what ABI should I target for PyPy?

I saw in [this pretty old thread](https://github.com/pyca/cryptography/issues/5696#issuecomment-763132517) that the `pypy73` ABI is suggested to be pretty stable. If I target that ABI, will I be able to create a "universal" wheel (per OS, obviously) that supports PyPy 3.10 7.3.X, *and* PyPy 3.9 7.3.X, *and* PyPy 3.8 7.3.X, etc?

If so, what should I tag that wheel as? `pp3-pypy73-{tag.platform}`? `py3-pypy73-{tag.platform}`? Something else?

**Comment by James-E-A** [comment]
If my understanding's incorrect, and PyPy is "actually supposed to" support ABI3 wheels, then I instead ask for suggestions on operating `setuptools` to build `py3-abi3-*` wheels that PyPy will recognize.

**Comment by mattip** [comment]
PyPy does not support the abi3 standard and cannot import abi3 wheels. See pypy/pypy#3397 for discussions about this: the conclusion was we need our own stable standard. Issue pypy/pypy#2617 explores the idea of having a shim to support the CPython C-API, but we would need to be very careful where the header files use macros. 

Eventually we hope [the HPy project](https://hpyproject.org/) will provide the answer for wheels that support any version and any implementation of python.

**Reply by James-E-A** [reply]
Thanks for the confirmation and the extra context.

So, for now, is `pp3-pypy73-{tag.platform}` the best target for PyPy? It's possible to build one wheel \(per platform\) that will work with PyPy3.8 through PyPy3.10?

**Reply by mattip** [reply, marked as answer]
> It's possible to build one wheel (per platform) that will work with PyPy3.8 through PyPy3.10?

No. One wheel per platform + pypy version
